src/org/exist/indexing/IndexController.java

```diff
@@ -202,22 +208,33 @@ public void removeCollection(final Collection collection, final DBBroker broker,
     /**
      * Re-index all nodes below the specified root node, using the given mode.
      *
-     * @param transaction the current transaction
+     * @param txn the current transaction
      * @param reindexRoot the node from which reindexing should occur
      * @param mode the mode, one of {@link ReindexMode#UNKNOWN}, {@link ReindexMode#STORE},
      * {@link ReindexMode#REMOVE_SOME_NODES} or {@link ReindexMode#REMOVE_ALL_NODES}.
      */
-    public void reindex(final Txn transaction, IStoredNode<? extends IStoredNode> reindexRoot, final ReindexMode mode) {
+    public void reindex(final Txn txn, final IStoredNode<? extends IStoredNode> reindexRoot, final ReindexMode mode) {
         if (reindexRoot == null) {
             return;
         }
+        IStoredNode<? extends IStoredNode> node = broker.objectWith(new NodeProxy(reindexRoot.getOwnerDocument(), reindexRoot.getNodeId()));
+
         setReindexing(true);
-        reindexRoot = broker.objectWith(new NodeProxy(reindexRoot.getOwnerDocument(), reindexRoot.getNodeId()));
-        setDocument(reindexRoot.getOwnerDocument(), mode);
-        getStreamListener();
-        IndexUtils.scanNode(broker, transaction, reindexRoot, listener);
-        flush();
-        setReindexing(false);
+        try {
+            StreamListener stream = getStreamListener(node.getOwnerDocument(), mode);
+            try {
+                stream.startIndexDocument(txn);
+
+                IndexUtils.scanNode(broker, txn, node, listener);
+            } finally {
+                stream.endIndexDocument(txn);
+            }
+
+            //that must be called outside this method to avoid slowdown
```

Comment:
That method to "internal" use, called from `ElementImpl` only. Also flush is happening at sync event on transaction release (and that only place where sync/flush must happen to avoid slowdown). I can't make this method protected because of poor packages design. Do you feel that I still should at comment to fava doc?